man/liffix.1
.TH liffix 1 03-November-2024 "LIF Utilities" "LIF Utilities"
.SH NAME
liffix \- fixes the medium-size information in the volume header of a LIF image file
.SH SYNOPSIS
.B lifutils liffix
\-m 
.I MEDIUMTYPE
[-p]
.I LIFIMAGE
.PP
.B lifutils liffix \-?
.SH DESCRIPTION
The volume header of the LIF file system contains information about the size of the volume specified by the numbers of tracks per surface, number of surfaces, and number of blocks per track. Some HP-IL controllers do not initialize these values correctly, which might worry software that works with LIF image files.  The medium size of the LIF image file can be checked with the
.I lifstat
program.
.B liffix
repairs the volume header of a LIF image file specified by the command line parameter 
.I LIFIMAGE.
.SH OPTIONS
.TP
.I \-m
This mandatory option specifies the medium type. The following medium types
are supported:
.RS
cass:    Initializes a medium with the size of an HP82161A cassette
.RE
.RS
disk:    Initializes a medium with the size of an HP9114B double-sided disk
.RE
.RS
hdrive1: Initializes a medium with the size of a 640 KB disk
.RE
.RS
hdrive2: Initializes a medium with the size of a 2 MB disk
.RE
.RS
hdrive4: Initializes a medium with the size of a 4 MB disk
.RE
.RS
hdrive8: Initializes a medium with the size of an 8 MB disk
.RE
.RS
hdrive16: Initializes a medium with the size of a 16 MB disk
.RE
.TP
.I \-p
Fix a LIF file system on a floppy disk. Note: this option is only supported on LINUX. Specify the floppy device instead of the lif-image-filename. See the LIFUTILS tutorial for details.
.TP
.I \-?
Print a message giving the program usage to standard error.
.SH EXAMPLES
If 
.I disk1.dat
is a LIF image file with improper medium-size information, then
.PP
.B lifutils liffix -m disk disk1.dat
.PP
will fix the medium-size information of
.I disk1.dat
The header is fixed with the parameters of the HP9114B double-sided disk
medium.
.PP
Note: A check is performed if the content of the image file matches the
size of the medium.
.SH REFERENCES
The LIF disk format is documented in the
.I HP\-IL Interface Owners Manual for the HP\-71 (Hewlett\-Packard)
, with further details (particularly HP41 and HP75 file types) in the 
.I HP\-41 Synthetic Quick Reference Guide (Jeremy Smith)
.SH AUTHOR
.B liffix
was written by Joachim Siebold, bug400@gmx.de  using code from Tony Duell, 
ard@p850ug1.demon.co.uk and has been placed under the GNU Public 
License version 2.0.


NEWS
change log information is in debian/changelog


src/lib/liffix.c
/* liffix.c -- Repair an illegal formatted LIF image file */
/* 2014 J. Siebold, and placed under the GPL */

#include <stdio.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <sys/stat.h>
#include "config.h"
#include "lifutils.h"
#include "lif_block.h"
#include "lif_dir_utils.h"
#include "lif_const.h"
#include "lif_create_entry.h"


#define DEBUG 0


void liffix_usage(void)
  {
    fprintf(stderr,"Usage : lifutils liffix -m Mediumtype [-p] LIFIMAGEFILE\n");
    fprintf(stderr,"        Fix the medium size information of a LIF image file\n");
    fprintf(stderr,"        -m Medium type (cass | disk | hdrive1/2/4/8/16) \n");
    fprintf(stderr,"        -p Fix a LIF file system on a floppy disk.\n");
    fprintf(stderr,"           Note: this option is only supported on LINUX.\n");
    fprintf(stderr,"           Specify the floppy device instead of the lif-image-filename.\n");
    fprintf(stderr,"           See the LIFUTILS tutorial for details.\n");
    fprintf(stderr,"\n");
  }

int liffix(int argc, char **argv)
  {
    /* System variables */
    int option; /* Command line option character */
    int lif_device; /* Descriptor of input device */
    int physical_flag; /* Option to use a physical device */
    char *medium= (char *) NULL;
    int totalblocks;
    int dir_length;
    int dir_block;
    int start_block;
    int num_blocks;
    int end_block;
    int maxblock;
    int dir_entry;
    int dir_end;
    int dir_start;
    int file_type;
    
    /* LIF disk values */
    unsigned char sector_data[SECTOR_SIZE]; /*  start sector */
    unsigned char dir_data[SECTOR_SIZE]; /* dir sector */

    /* Process command line options */
    optind=1;
    physical_flag=0;
    while ((option=getopt(argc,argv,"pm:x?"))!=-1)
      {
        switch(option)
          {
            case 'p' : physical_flag=1;
                        break;
            case 'm':  medium=optarg;
                       break;
            case '?' : liffix_usage();
                       return(RETURN_OK);
          }
      }

    /* Are the right number of names specified ? */
    if( optind != argc-1  ) 
      {
        /* No, give an error */
        liffix_usage();
        return(RETURN_ERROR);
      }
    /* Is the medium type specified ? */
    if(medium == (char *) NULL)
      {
        liffix_usage();
        return(RETURN_ERROR);
      }
    /* Open lif device */
    if((lif_device=lif_open(argv[optind],O_RDWR | O_BINARY,0,physical_flag))==-1)
      {
        fprintf(stderr,"Error opening %s\n",argv[optind]);
        return(RETURN_ERROR);
      }
    /* Now read block 0 to find where the directory is */
    lif_read_block(lif_device,0,sector_data);

    /* Make sure it's a LIF disk */
    if(get_lif_int(sector_data+0,2)!=0x8000)
      {
        fprintf(stderr,"This is not a LIF disk!\n");
        return(RETURN_ERROR);
      }
    /* Find where the directory is */
    dir_start=get_lif_int((sector_data+8),4);
    dir_length=get_lif_int((sector_data+16),4);

    /* Now scan the directory to find the last block used by the filesystem */
    maxblock=0;
    for(dir_block=0; dir_block<dir_length; dir_block++)
      {
         dir_end=0;
         lif_read_block(lif_device,dir_block+dir_start,dir_data);
         for(dir_entry=0; dir_entry<8; dir_entry++)
           {
             file_type=get_lif_int((dir_data+(dir_entry<<5)+10),2);
             if(file_type==0) { continue; } /* Skip over deleted files */
             if(file_type==0xffff)
               {
                 /* This is the end of the directory */
                 dir_end=1;
                 break;
               }
             start_block=get_lif_int(dir_data+(dir_entry<<5)+12,4);
             num_blocks=get_lif_int(dir_data+(dir_entry<<5)+16,4);
             end_block= start_block+num_blocks;
             if(end_block > maxblock) maxblock= end_block;
           }
         if(dir_end) { break; } /* Quit at end of directory */
      }

    /* Tracks, heads, sectors */
    totalblocks=0;
    if(strcmp(medium,"cass")==0) {
        put_lif_int(sector_data+24,4,2);
        put_lif_int(sector_data+28,4,1);
        put_lif_int(sector_data+32,4,256);
        if ((sector_data[18] & 0x40) != 0) sector_data[18] &= ~ 0x40;
        *(sector_data+18)=0;
        totalblocks= 512;
    }
    if(strcmp(medium,"disk")==0) {
        put_lif_int(sector_data+24,4,77);
        put_lif_int(sector_data+28,4,2);
        put_lif_int(sector_data+32,4,16);
        if ((sector_data[18] & 0x40) != 0) sector_data[18] &= ~ 0x40;
        totalblocks=2464; /* 16 spare blocks */
    }
    if(strcmp(medium,"hdrive1")==0) {
        put_lif_int(sector_data+24,4,80);
        put_lif_int(sector_data+28,4,2);
        put_lif_int(sector_data+32,4,16);
        *(sector_data+18)=0;
        totalblocks=2560; 
    }
    if(strcmp(medium,"hdrive2")==0) {
        put_lif_int(sector_data+24,4,125);
        put_lif_int(sector_data+28,4,1);
        put_lif_int(sector_data+32,4,64);
        totalblocks=8000; 
    }
    if(strcmp(medium,"hdrive4")==0) {
        put_lif_int(sector_data+24,4,125);
        put_lif_int(sector_data+28,4,2);
        put_lif_int(sector_data+32,4,64);
        totalblocks=16000; 
    }
    if(strcmp(medium,"hdrive8")==0) {
        put_lif_int(sector_data+24,4,125);
        put_lif_int(sector_data+28,4,4);
        put_lif_int(sector_data+32,4,64);
        totalblocks=32000; 
    }
    if(strcmp(medium,"hdrive16")==0) {
        put_lif_int(sector_data+24,4,125);
        put_lif_int(sector_data+28,4,8);
        put_lif_int(sector_data+32,4,64);
        totalblocks=64000; 
    }
    if(totalblocks==0) {
       liffix_usage();
        return(RETURN_ERROR);
      }
    debug_print("Last block used %d\n",maxblock);
    debug_print("Total blocks of medium %d\n",totalblocks);
    if(totalblocks < maxblock) 
     {
        fprintf(stderr,"Medium too small\n");
        lif_close(lif_device);
        return(RETURN_ERROR);
      }
    /* set LIF Version number */
    put_lif_int(sector_data+20,2,1);

    /* Now write block 0  */
    lif_write_block(lif_device,0,sector_data);
    /* tidy up and quit */
    lif_close(lif_device);
    return(RETURN_OK);
  }
